pkg/controller/controllerutil/controllerutil.go

```diff
@@ -97,3 +101,78 @@ func referSameObject(a, b v1.OwnerReference) bool {
 
 	return aGV == bGV && a.Kind == b.Kind && a.Name == b.Name
 }
+
+// OperationType is the action result of a CreateOrUpdate call
+type OperationType string
+
+const ( // They should complete the sentence "Deployment default/foo has been ..."
+	// OperationNoop means that the resource has not been changed
+	OperationNoop = "unchanged"
+	// OperationCreate means that a new resource is created
+	OperationCreate = "created"
+	// OperationUpdate means that an existing resource is updated
+	OperationUpdate = "updated"
+)
+
+// CreateOrUpdate the given object obj in the Kubernetes cluster. The object's
+// desired state should be reconciled with the existing state using the passed
+// in ReconcileFn. obj must be a struct pointer so that obj can be updated with
+// the content returned by the Server.
+//
+// It returns the executed operation and an error.
+func CreateOrUpdate(ctx context.Context, c client.Client, obj runtime.Object, r ReconcileFn) (OperationType, error) {
+	// op is the operation we are going to attempt
+	var op OperationType = OperationNoop
+
+	// get the existing object meta
+	metaObj, ok := obj.(v1.Object)
+	if !ok {
+		return OperationNoop, fmt.Errorf("%T does not implement metav1.Object interface", obj)
+	}
+
+	// retrieve the existing object
+	key := client.ObjectKey{
+		Name:      metaObj.GetName(),
+		Namespace: metaObj.GetNamespace(),
+	}
+	err := c.Get(ctx, key, obj)
+
+	// reconcile the existing object
+	existing := obj.DeepCopyObject()
+	existingObjMeta := existing.(v1.Object)
+	existingObjMeta.SetName(metaObj.GetName())
+	existingObjMeta.SetNamespace(metaObj.GetNamespace())
+
+	if e := r(obj); e != nil {
+		return OperationNoop, e
+	}
+
+	if metaObj.GetName() != existingObjMeta.GetName() {
+		return OperationNoop, fmt.Errorf("ReconcileFn cannot mutate objects name")
+	}
+
+	if metaObj.GetNamespace() != existingObjMeta.GetNamespace() {
+		return OperationNoop, fmt.Errorf("ReconcileFn cannot mutate objects namespace")
+	}
+
+	if errors.IsNotFound(err) {
+		err = c.Create(ctx, obj)
+		op = OperationCreate
+	} else if err == nil {
+		if reflect.DeepEqual(existing, obj) {
+			return OperationNoop, nil
+		}
+		err = c.Update(ctx, obj)
```

Comment:
Trying to analyze the semantics of this wrt status sub-resource. If an API implements `status` as sub-resource (which is the recommended approach), then `Update` is updating just the `Spec` part of API (Then the `DeepEqual` should be of the spec's field only ?) @DirectXMan12

Reply:
@droot  `obj` being a `runtime.Object`, is it possible to get the spec of that? If so what should happen with types not having a `Spec` field?